upf/upf_api.c

```diff
@@ -555,6 +554,161 @@ vl_api_upf_nwi_dump_t_handler (vl_api_upf_nwi_dump_t * mp)
   }
 }
 
+/* API message handler */
+static void
+vl_api_upf_pfcp_endpoint_add_del_t_handler (vl_api_upf_pfcp_endpoint_add_del_t
+					    * mp)
+{
+  vl_api_upf_pfcp_endpoint_add_del_reply_t *rmp = NULL;
+  upf_main_t *sm = &upf_main;
+  u32 vrf;
+  u32 fib_index = 0;
+  u8 is_add;
+  ip46_address_t ip_addr;
+  int rv = 0;
+
+  is_add = mp->is_add;
+  vrf = clib_net_to_host_u32 (mp->table_id);
+
+  ip_address_decode (&mp->ip, &ip_addr);
+  if (ip46_address_is_zero (&ip_addr))
+    {
+      rv = VNET_API_ERROR_INVALID_VALUE;
+      upf_debug ("IP should be provided");
+      goto out;
+    }
+
+  if (vrf != ~0)
+    {
+      fib_index =
+	fib_table_find (fib_ip_proto (!ip46_address_is_ip4 (&ip_addr)), vrf);
+      if (fib_index == ~0)
+	{
+	  rv = VNET_API_ERROR_NO_SUCH_TABLE;
+	  upf_debug ("nonexistent vrf %d", vrf);
+	  goto out;
+	}
+    }
+
+  rv = vnet_upf_pfcp_endpoint_add_del (&ip_addr, fib_index, is_add);
+
+out:
+
+  REPLY_MACRO (VL_API_UPF_PFCP_ENDPOINT_ADD_DEL_REPLY);
+}
+
+static void
+send_upf_pfcp_endpoint_details (vl_api_registration_t * reg,
+				ip46_address_fib_t * key, u32 context)
+{
+  vl_api_upf_pfcp_endpoint_details_t *mp;
+  upf_main_t *sm = &upf_main;
+
+  mp = vl_msg_api_alloc (sizeof (*mp));
+  clib_memset (mp, 0, sizeof (*mp));
+
+  mp->_vl_msg_id = htons (VL_API_UPF_PFCP_ENDPOINT_DETAILS + sm->msg_id_base);
+  mp->context = context;
+
+  mp->fib_table = htonl (key->fib_index);
+  ip_address_encode (&key->addr, IP46_TYPE_ANY, &mp->ip);
+
+  vl_api_send_msg (reg, (u8 *) mp);
+}
+
+/* API message handler */
+static void
+vl_api_upf_pfcp_endpoint_dump_t_handler (vl_api_upf_pfcp_endpoint_dump_t * mp)
+{
+  upf_main_t *sm = &upf_main;
+  vl_api_registration_t *reg;
+  ip46_address_fib_t *key;
+  uword *v;
+
+  reg = vl_api_client_index_to_registration (mp->client_index);
+  if (!reg)
+    {
+      return;
+    }
+  /* *INDENT-OFF* */
+  mhash_foreach(key, v, &sm->pfcp_endpoint_index,
+  ({
+    send_upf_pfcp_endpoint_details (reg, key, mp->context);
+  }));
+  /* *INDENT-ON* */
+}
+
+/* API message handler */
+static void
+vl_api_upf_pfcp_server_set_t_handler (vl_api_upf_pfcp_server_set_t * mp)
+{
+  vl_api_upf_pfcp_server_set_reply_t *rmp = NULL;
+  upf_main_t *sm = &upf_main;
+  u32 fifo_size = 0;
+  u32 prealloc_fifos = 0;
+  u64 segment_size = 0;
+  int rv = 0;
+
+  /* We get segment size in MB */
+  segment_size = (u64) clib_net_to_host_u32 (mp->segment_size);
+  segment_size <<= 20;
+
+  if (segment_size >= 0x100000000ULL)
```

Comment:
It would be better to check before shifting the segment_size. Otherwise, very large values would be changed to zero instead of triggering an error.

Reply:
Done